src/components/ReactFlowBox.tsx

```diff
@@ -20,126 +21,192 @@ import { EdgeData, NodeData, NodeSchema } from '../common-types';
-    const [_edges, _setEdges, onEdgesChange] = useEdgesState<EdgeData>([]);
-
-    const sortNodesAndEdges = useCallback(() => {
-        const iterators = nodes.filter((n) => n.type === 'iterator'); // .sort((i) => (i.selected ? 1 : -1));
-        let sortedNodes: Node<NodeData>[] = [];
-
-        // Sort the nodes in a way that makes iterators stack on each other correctly
-        // Put iterators below their children
-        iterators.forEach((_iterator, index) => {
-            const iterator = _iterator;
-            iterator.zIndex = STARTING_Z_INDEX + index * 5;
-            sortedNodes.push(iterator);
-            const children = nodes.filter((n) => n.parentNode === iterator.id);
-            // sorted.concat(children);
-            children.forEach((_child) => {
-                const child = _child;
-                child.zIndex = STARTING_Z_INDEX + index * 5 + 1;
-                // child.position.x = Math.min(Math.max(child.position.x, 0), iterator.width);
-                // child.position.y = Math.min(Math.max(child.position.y, 0), iterator.height);
-                sortedNodes.push(child);
-            });
-        });
-
-        // Put nodes not in iterators on top of the iterators
-        const freeNodes = nodes.filter((n) => n.type !== 'iterator' && !n.parentNode);
-        freeNodes.forEach((f) => {
-            sortedNodes.push(f);
-        });
-
-        const indexedEdges = edges.map((e) => {
-            const index = sortedNodes.find((n) => n.id === e.target)?.zIndex || 1000;
-            return { ...e, zIndex: index };
-        });
-
-        // This fixes the connection line being behind iterators if no edges are present
-        if (indexedEdges.length === 0) {
-            sortedNodes = sortedNodes.map((n) => ({ ...n, zIndex: -1 }));
-        }
+    const reactFlowInstance = useReactFlow();
 
-        _setNodes(sortedNodes);
-        _setEdges(indexedEdges);
-    }, [nodes, edges, _setNodes, _setEdges]);
+    const [nodes, setNodes, onNodesChange] = useNodesState<NodeData>([]);
+    const [edges, setEdges, onEdgesChange] = useEdgesState<EdgeData>([]);
 
     useEffect(() => {
-        sortNodesAndEdges();
-    }, [nodes, edges]);
+        setSetNodes(() => setNodes);
+        setSetEdges(() => setEdges);
+    }, [setNodes, setEdges]);
+
+    const selectedNodesKey = useMemo(
+        () =>
+            nodes
+                .filter((n) => n.selected)
+                .map((n) => n.id)
+                .join(''),
+        [nodes]
+    );
+    const selectedEdgesKey = useMemo(
+        () =>
+            edges
+                .filter((e) => e.selected)
+                .map((e) => e.id)
+                .join(''),
+        [edges]
+    );
+
+    // We don't want this to cause a re-render, so we will commit the greatest sin
+    // known to React developers: interior mutation.
+    useMemo(() => {
+        nodes.sort((a, b) => a.id.localeCompare(b.id));
```

Comment:
We get determinism. The z order of components is determined by their respective z indexes and their order in the DOM. By sorting the array by id (actually any total order would do, but I chose id because it's easy), we ensure that 2 nodes with the same z index always have the same z order. This means that overlapping nodes won't jump around over each other.

Reply:
Sounds good